### ISSUE TITLE
Pylance reports false errors then crashes

### ISSUE THREAD

**Issue by paul-rogers** [microsoft/pylance-release#6178]
Type: <b>Bug</b>

Use a Python notebook. Type some code. At some point, I'll get something like this:

```text
[Cell 1]:
def foo():
    return 1

[Cell 2]:
x = foo()
```

Pylance will report the reference to `foo`, in Cell 2, as undefined (that is, with a yellow squiggly line.) This only occurs part of the time. I've not found a reliable reproduction sequence; the error seems to be random and occurs on a small subset of symbols. The symbols are, in fact, correct: the code in Cell 2 executes properly.

Once the symbol is marked undefined, there seems to be no way to convince Pylance that it is defined. Deleting and re-adding Cell 1 does not work. Renaming `foo` to `bar` and back doesn't work.

This will continue for a while, then Pylance will crash as indicated by an alert in the lower-right corner of the screen: a dialog when then offers to run Pylance in a node executable. The latest crash seems to have been triggered when I cut "using 'x'" a cell from the notebook; a cell which contained Python code.

The only workaround appears to be to quit VS Code and restart.

This issue has appeared only with the latest update: there were no such problems in any of the previous versions over the last year or so.

Extension version: 2024.7.1
VS Code version: Code 1.91.1 (f1e16e1e6214d7c44d078b1f0607b2388f29d729, 2024-07-09T22:08:12.169Z)
OS version: Linux x64 5.15.0-113-generic
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-2600K CPU @ 3.40GHz (8 x 531)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|2, 2, 2|
|Memory (System)|31.31GB (8.93GB free)|
|Process Argv|--crash-reporter-id 5e232e7b-355a-468c-ac4a-02c300f898bd|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|cinnamon|
|XDG_CURRENT_DESKTOP|X-Cinnamon|
|XDG_SESSION_DESKTOP|cinnamon|
|XDG_SESSION_TYPE|x11|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
vstes627:30244334
vscorecescf:30445987
vscod805:30301674
binariesv615:30325510
vsaa593:30376534
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
a9j8j154:30646983
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonnoceb:30805159
asynctok:30898717
pythonregdiag2:30936856
pythonmypyd1:30879173
h48ei257:31000450
pythontbext0:30879054
accentitlementsc:30995553
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
chatpanelt:31048053
dsvsc021:30996838
g316j359:31013175
pythoncenvpt:31062603
a69g1124:31058053
dvdeprecation:31068756
dwnewjupytercf:31046870
2f103344:31071589
impr_priority:31094925
pythonrstrctxtcf:31093870

```

</details>

**Comment by debonte** [comment]
Can you please provide a log as described under "Filing an issue" in [the troubleshooting guide](https://github.com/microsoft/pylance-release/blob/main/TROUBLESHOOTING.md#filing-an-issue)?

**Comment by paul-rogers** [comment]
@debonte, I've turned on logging. The problem occurs infrequently: I'll post the results when the problem next occurs.

**Comment by paul-rogers** [comment]
@debonte, Here is one way to trigger the error in which Pylance reports a symbol as undefined when it is, in fact, defined. This scenario also reproduces the crash.

1. Create a new Notebook.
2. Define a Python cell:
```python
def foo():
    pass
```
3. Define a second Python cell below the first:
```python
foo()
```
4. Go back to the first cell: use 'x' to cut that cell. Notice that the reference `foo()` in the second cell becomes undefined.
5. Create a new Python cell above the reference cell. Retype the trivial function declaration.
6. Notice that the reference in the second cell, `foo()`, is still marked as undefined. 

It now becomes hard to correct the error. Here are some things to try:

1. Rename `def foo():` to `def foo2():` by typing the "2". Then, remove the "2" to get `foo()` again. The yellow squiggle on the reference does not disappear (but it should).
2. Do the same operation, but use the Rename (F2) operation. The reference is not renamed.
3. Rename `def foo():` to `def bar():` by typing the new name. Type the new name for the reference also. We've now got an entirely new symbol. However, the reference to new symbol (`bar`) is still marked undefined.
4. Create two entirely new cells, at the end of the notebook, but this time use the symbol name `mumble`. The reference is correctly displayed without the yellow squiggle.
5. Drag the two new cells to the top of the notebook. You how have `def mumble...`, `mumble()`, `def foo...`, `foo()`. Edit `foo()` to read `mumble()`. You will now have one `mumble()` reference without a squiggle, this edited one with a squiggle.

From the above, it appears that cutting a definition (using 'x' in a notebook: remember we cut `def foo():` above), somehow places the reference cells in a bad state. Such cells seem  to be not checked on future edits.

The logs were requested. I enabled logging as per the instructions given in the link. However, there is nothing in the output that suggests anything is amiss: the output lines look basically identical when things work (before the cut operation) and afterward.

Now, let's get a crash:

1. Select the top cell, the one with `def mumble()...`. Cut it with 'x'. The reference is now marked undefined, as expected.
2. Undo the cut with `Edit` &rarr; `Undo`.
3. We get the dialog `Pylance has crashed...`.
4. Contents of the `Output` window:

```text                                                                                                                                                                                                            ^
Error: Debug Failure. False expression: Chained file path file:///home/paul/git/play/Test4.ipynb.py#W2sZmlsZQ%3D%3D doesn't match cellFilePaths file:///home/paul/git/play/Test4.ipynb.py#W4sZmlsZQ%3D%3D
    at _0x2ee97f (/home/paul/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/server.bundle.js:1:888452)
    at _0x133320 (/home/paul/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/server.bundle.js:1:892655)
    at _0x48f037 (/home/paul/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/server.bundle.js:1:893099)
    at /home/paul/.vscode/extensions/ms-python.vscode-pylance-2024.7.1/dist/server.bundle.js:1:819832

2024-07-26 18:29:27.857 [info] [Error - 6:29:27 PM] Server process exited with code 1.
2024-07-26 18:29:27.908 [info] [Error - 6:29:27 PM] Request textDocument/foldingRange failed.
2024-07-26 18:29:27.908 [info] Error: Connection is disposed.
	at de (/home/paul/.vscode/extensions/ms-python.python-2024.10.0-linux-x64/out/client/extension.js:2:1948291)
	at Object.sendRequest (/home/paul/.vscode/extensions/ms-python.python-2024.10.0-linux-x64/out/client/extension.js:2:1950595)
	at S.sendRequest (/home/paul/.vscode/extensions/ms-python.python-2024.10.0-linux-x64/out/client/extension.js:2:1991691)
	at async Z.provideFoldingRanges (/usr/share/code/resources/app/out/vs/workbench/api/node/extensionHostProcess.js:154:126701)
2024-07-26 18:29:29.253 [info] [Error - 6:29:29 PM] Notify file events failed.
2024-07-26 18:29:29.254 [info] Error: Connection is disposed.
	at de (/home/paul/.vscode/extensions/ms-python.python-2024.10.0-linux-x64/out/client/extension.js:2:1948291)
	at Object.sendNotification (/home/paul/.vscode/extensions/ms-python.python-2024.10.0-linux-x64/out/client/extension.js:2:1949134)
	at S.sendNotification (/home/paul/.vscode/extensions/ms-python.python-2024.10.0-linux-x64/out/client/extension.js:2:1992950)
	at async t.Delayer.<anonymous> (/home/paul/.vscode/extensions/ms-python.python-2024.10.0-linux-x64/out/client/extension.js:2:2004892)
```

Before the lines above, there is a vast amount of text that looks like code or debug output: far too much material to paste here.

Side note: Clicking the `x` close button on the `Pylance has crashed...` dialog closes the dialog, but it immediately opens again. This happens over and over. One has to click on `No` to get the dialog to actually disappear.

Clicking `No` appears to restart Pylance as a large amount of output appears. After this, the references in the notebook are correctly resolved and the yellow squiggles disappear.

**Comment by paul-rogers** [comment]
Another path to an incorrectly labeled undefined symbol.

1. Create a notebook cell: `pprint(10)`. `pprint` will be marked undefined.
2. Create a cell above: `from pprint import pprint`. `pprint` is now defined, but the first cell how has the cell marked as undefined.
3. Execute the first cell: no change in the reference.
4. Execute the second cell. It works, showing that `pprint` is, in fact, defined.
5. Create a third cell, below the other two: `pprint(20)`. Now, `pprint` is not marked as undefined.

At this point, we have one reference to `pprint` marked with a yellow squiggle (undefined) and a second one without the squiggle (defined). So, a wild guess is that, in this case, adding the cell with the import put the existing reference cell in a bad state.

FWIW, here are typical output entries for this scenario:

```text
2024-07-26 19:15:56.537 [info] (2610877) [BG(3)] SemanticTokens delta previousResultId:1722046556301 at file:///home/paul/git/play/Test4.ipynb.py#W6sZmlsZQ%3D%3D (0ms)
2024-07-26 19:15:57.390 [info] [Info  - 7:15:57 PM] (2610877) SourceFile: Received fs event 'change' for path '/home/paul/git/play/Test4.ipynb'
```

**Comment by heejaechang** [comment]
thank you for the repro, I can repro this.

**Comment by heejaechang** [comment]
so, I took a look. looks like LSP doesn't raise `notebook change` event for cell ordering so we don't know whether cell ordering has changed.

**Comment by heejaechang** [comment]
opened issue on https://github.com/microsoft/vscode-languageserver-node/issues/1525

we probably will rollback to old version of lsp packages until vscode-languageserver-node release new packages with the issue fixed.

**Comment by paul-rogers** [comment]
@heejaechang, thanks for tracking down the issue!

**Comment by heejaechang** [comment]
This issue has been fixed in prerelease version 2024.7.104, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#20247104-31-July-2024-PreRelease)
